### ISSUE TITLE
🐛 Bug Report: Unable to publish template to Azure

### ISSUE THREAD

**Issue by mattemoore** [backstage/backstage#26368]
### 📜 Description

This bug was noted here but case was closed as it went stale: https://github.com/backstage/backstage/issues/13938

The publish:azure template action fails when pushing the code to Azure Git Repo when using Managed Identity for authentication.

We are using managed identity to integrate Backstage with Azure DevOps.

1. We can read from repo from backstage using MI.
2. We can create a repo from the template using MI.
3. We CANNOT push the initial commit to the repo from the template using MI.
4. NOTE: We CAN push the initial commit to the repo from the template when running local dev which uses a Personal Access Token with Read/Write/Manage Code scope:

`app-config.production.yaml`:
```
# In production environment, Backstage Azure DevOps plugin (README, CI/CD) uses Managed Identity to query data in Azure DevOps
# NOTE: Currently the Backstage Azure Devops Plugin is in a transition from `azureDevOps` key used for auth info to using `integrations.azure`
#       During this transition the `host` and `organization` settings are pulled from `azureDevOps` and the `credentials` are pulled from `integrations.azure`
#       Eventually the `azureDevOps` config key will be removed entirely.
azureDevOps:
  host: dev.azure.com
  token: dummy-token-value
  organization: REDACTED

# In production environment, Backstage Catalog plugin uses Managed Identity in App Service to load catalog entities from Azure DevOps Repositories
integrations:
  azure:
    - host: dev.azure.com
      credentials:
        - clientId: ${BACKSTAGE_MANAGED_IDENTITY_CLIENT_ID}
       # NOTE: PAT token from local dev machine works
       # - personalAccessToken: ${AZURE_DEVOPS_PAT} # This works
```

```
[69](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-69)2024-08-29T18:05:14.755Z info: Init git repository {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30}

[70](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-70)2024-08-29T18:05:14.933Z info: Adding file {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,filepath=.}

[71](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-71)2024-08-29T18:05:17.484Z info: Committing file to repo {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,message=initial commit}

[72](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-72)2024-08-29T18:05:17.595Z info: Creating new remote {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,remote=origin,url=https://REDACTED@dev.azure.com/REDACTED/REDACTED/_git/foobar12345}

[73](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-73)2024-08-29T18:05:17.605Z info: Pushing directory to remote {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,remote=origin}

[74](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-74)2024-08-29T18:05:17.722Z error: Failed to push to repo {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30, remote=origin}

[75](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-75)2024-08-29T18:05:17.725Z Error: HTTP Error: 401 Unauthorized {data={"statusCode":401,"statusMessage":"Unauthorized","response":""}}

[76](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-76)    at Git.push (/app/node_modules/@backstage/plugin-scaffolder-node/dist/index.cjs.js:289:15)

[77](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-77)    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

[78](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-78)    at async Object.initRepoAndPush (/app/node_modules/@backstage/plugin-scaffolder-node/dist/index.cjs.js:377:3)

[79](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-79)
```

### 👍 Expected behavior

After creating new repository the skeleton code should be pushed to the target Git Repo.

### 👎 Actual Behavior with Screenshots

After creating the new repository the git push error fails with 401.  

```
[69](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-69)2024-08-29T18:05:14.755Z info: Init git repository {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30}

[70](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-70)2024-08-29T18:05:14.933Z info: Adding file {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,filepath=.}

[71](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-71)2024-08-29T18:05:17.484Z info: Committing file to repo {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,message=initial commit}

[72](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-72)2024-08-29T18:05:17.595Z info: Creating new remote {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,remote=origin,url=https://REDACTED@dev.azure.com/REDACTED/REDACTED/_git/foobar12345}

[73](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-73)2024-08-29T18:05:17.605Z info: Pushing directory to remote {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30,remote=origin}

[74](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-74)2024-08-29T18:05:17.722Z error: Failed to push to repo {dir=/tmp/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30, remote=origin}

[75](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-75)2024-08-29T18:05:17.725Z Error: HTTP Error: 401 Unauthorized {data={"statusCode":401,"statusMessage":"Unauthorized","response":""}}

[76](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-76)    at Git.push (/app/node_modules/@backstage/plugin-scaffolder-node/dist/index.cjs.js:289:15)

[77](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-77)    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

[78](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-78)    at async Object.initRepoAndPush (/app/node_modules/@backstage/plugin-scaffolder-node/dist/index.cjs.js:377:3)

[79](https://developers.dev.tc.gc.ca/create/tasks/7cce4f9f-18db-41d7-a822-5b7a2e3e3e30#line-79)
```

### 👟 Reproduction steps

`app-config.production.yaml`:
```
# In production environment, Backstage Azure DevOps plugin (README, CI/CD) uses Managed Identity to query data in Azure DevOps
# NOTE: Currently the Backstage Azure Devops Plugin is in a transition from `azureDevOps` key used for auth info to using `integrations.azure`
#       During this transition the `host` and `organization` settings are pulled from `azureDevOps` and the `credentials` are pulled from `integrations.azure`
#       Eventually the `azureDevOps` config key will be removed entirely.
azureDevOps:
  host: dev.azure.com
  token: dummy-token-value
  organization: REDACTED

# In production environment, Backstage Catalog plugin uses Managed Identity in App Service to load catalog entities from Azure DevOps Repositories
integrations:
  azure:
    - host: dev.azure.com
      credentials:
        - clientId: ${BACKSTAGE_MANAGED_IDENTITY_CLIENT_ID}
       # NOTE: PAT token from local dev machine works
       # - personalAccessToken: ${AZURE_DEVOPS_PAT} # This works
```

`template.yaml`:
```
apiVersion: scaffolder.backstage.io/v1beta3
# https://backstage.io/docs/features/software-catalog/descriptor-format#kind-template
kind: Template
metadata:
  name: aspnetcore8mvc
  title: ASP.NET core 8 MVC Web Application w/ All Toppings
  description: Foo Description

  owner: group:technical-advisor-center-of-excellence
  type: service

  # These parameters are used to generate the input form in the frontend, and are
  # used to gather input data for the execution of the template.
  parameters:
    - title: Provide name and owner of new component
      required:
        - name
      properties:
        name:
          title: Name
          type: string
          description: Unique name of the component to be created.  This name will be used in the Portal registration and the Pull Request created.
          ui:autofocus: true
          ui:options:
            rows: 5
        owner:
          title: Owner
          type: string
          description: Owner of the component
          ui:field: OwnerPicker
          ui:options:
            catalogFilter:
              kind: Group
    - title: Choose a target repo for the code generated
      required:
        - repoUrl
      properties:
        repoUrl:
          title: Repository Location
          type: string
          ui:field: RepoUrlPicker
          ui:options:
            allowedHosts:
              - dev.azure.com
            allowedOrganizations:
              - REDACTED
  steps:
    - id: fetch-base
      name: Fetch Base
      action: fetch:template
      input:
        url: ./skeleton
        values:
          component_id: ${{ parameters.name }}
          destination: ${{ parameters.repoUrl | parseRepoUrl }}
          owner: ${{ parameters.owner }}
          project: ${{ parameters.project }}
    - id: publish
      action: publish:azure
      name: Publish to Azure
      input:
        repoUrl: ${{ parameters.repoUrl }}
        defaultBranch: main
    - id: register
      name: Register
      action: catalog:register
      input:
        repoContentsUrl: ${{ steps['publish'].output.repoContentsUrl }}
        catalogInfoPath: '/catalog-info.yaml'
  output:
    links:
      - title: Repository
        url: ${{ steps['publish'].output.remoteUrl }}
      - title: Open in catalog
        icon: catalog
        entityRef: ${{ steps['register'].output.entityRef }}
```

### 📃 Provide the context for the Bug.

Trying to create templated apps/apis so we bring consistency to our org as well as enable changes to non-functional requirements and governance to just be adopted by dev teams without adding more cognitive load on them.

### 🖥️ Your Environment

yarn run v1.22.21
$ /REDACTED/node_modules/.bin/backstage-cli info
OS:   Linux 5.15.153.1-microsoft-standard-WSL2 - linux/x64
node: v20.9.0
yarn: 1.22.21
cli:  0.24.0 (installed)
backstage:  1.29.2

Dependencies:
  @backstage/app-defaults                                          1.5.9
  @backstage/backend-app-api                                       0.7.5, 0.8.0
  @backstage/backend-common                                        0.21.7, 0.22.0, 0.23.3
  @backstage/backend-defaults                                      0.4.1
  @backstage/backend-dev-utils                                     0.1.4
  @backstage/backend-openapi-utils                                 0.1.15
  @backstage/backend-plugin-api                                    0.6.18, 0.7.0
  @backstage/backend-tasks                                         0.5.27
  @backstage/catalog-client                                        1.6.5
  @backstage/catalog-model                                         1.5.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.26.11
  @backstage/config-loader                                         1.8.1
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.1
  @backstage/core-compat-api                                       0.2.7
  @backstage/core-components                                       0.14.9
  @backstage/core-plugin-api                                       1.9.3
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-plugin-api                                   0.6.7
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.29
  @backstage/integration                                           1.13.0
  @backstage/plugin-api-docs                                       0.11.7
  @backstage/plugin-app-backend                                    0.3.71
  @backstage/plugin-app-node                                       0.1.22
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.3
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.14
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.5
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.5
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.1.5
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.17
  @backstage/plugin-auth-backend-module-github-provider            0.1.19
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.19
  @backstage/plugin-auth-backend-module-google-provider            0.1.19
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.17
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.3
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.15
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.3
  @backstage/plugin-auth-backend-module-okta-provider              0.0.15
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.3
  @backstage/plugin-auth-backend                                   0.22.9
  @backstage/plugin-auth-node                                      0.4.17
  @backstage/plugin-auth-react                                     0.1.4
  @backstage/plugin-bitbucket-cloud-common                         0.2.21
  @backstage/plugin-catalog-backend-module-azure                   0.1.42
  @backstage/plugin-catalog-backend-module-logs                    0.0.1
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.20
  @backstage/plugin-catalog-backend                                1.24.0
  @backstage/plugin-catalog-common                                 1.0.25
  @backstage/plugin-catalog-graph                                  0.4.7
  @backstage/plugin-catalog-import                                 0.12.1
  @backstage/plugin-catalog-node                                   1.12.4
  @backstage/plugin-catalog-react                                  1.12.2
  @backstage/plugin-catalog                                        1.21.1
  @backstage/plugin-events-node                                    0.3.8
  @backstage/plugin-org                                            0.6.27
  @backstage/plugin-permission-backend                             0.5.46
  @backstage/plugin-permission-common                              0.7.13, 0.8.0
  @backstage/plugin-permission-node                                0.7.29, 0.8.0
  @backstage/plugin-permission-react                               0.4.24
  @backstage/plugin-proxy-backend                                  0.5.3
  @backstage/plugin-scaffolder-backend-module-azure                0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.12
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.12
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.12
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.14
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.12
  @backstage/plugin-scaffolder-backend-module-github               0.4.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.4
  @backstage/plugin-scaffolder-backend                             1.23.0
  @backstage/plugin-scaffolder-common                              1.5.4
  @backstage/plugin-scaffolder-node                                0.4.8
  @backstage/plugin-scaffolder-react                               1.10.0
  @backstage/plugin-scaffolder                                     1.23.0
  @backstage/plugin-search-backend-module-catalog                  0.1.28
  @backstage/plugin-search-backend-module-pg                       0.5.32
  @backstage/plugin-search-backend-module-techdocs                 0.1.27
  @backstage/plugin-search-backend-node                            1.2.27
  @backstage/plugin-search-backend                                 1.5.14
  @backstage/plugin-search-common                                  1.2.13
  @backstage/plugin-search-react                                   1.7.13
  @backstage/plugin-search                                         1.4.14
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-backend                               1.10.9
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.12
  @backstage/plugin-techdocs-node                                  1.12.8
  @backstage/plugin-techdocs-react                                 1.2.6
  @backstage/plugin-techdocs                                       1.10.7
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.10
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.9
  @backstage/theme                                                 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8
Done in 0.68s.

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by benjdlambert** [comment]
Hmm, is this something to do with `permissions` or is it something to do with how we setup username and password in the `Git` client when using MI? I'm not super familiar with the Azure toolset, so if we can get some more help here would be great :pray:

**Comment by mattemoore** [comment]
@benjdlambert I'm happy to help debug or test but not equipped to contribute to the azure scaffolder.

**Comment by tomdean22** [comment]
Same issue. The repos are being created, but the initial commit fails unless a personal access token is used, which isn't an option for our production environment. I'm using an app registration instead of a managed identity.

**Comment by benjdlambert** [comment]
@mattemoore @tomdean22 has this worked for either of you before or is this a new setup and it doesn't work?

@mattemoore if you can maybe have a play around with the above code in `node_modules` to see if you can get something different would be super useful. You could start by removing the `if` clause and always provide `username` and `password` which happen only when you provide a token and see if that yields different results.

**Comment by mattemoore** [comment]
@benjdlambert this is a new setup for us.  I can add that our Catalog backend integration uses Azure Managed Identity successfully.

@benjdlambert  I can poke around the code next week.  I will have to change to use app registration as managed identity won't work on my local dev environment.

**Comment by tomdean22** [comment]
> @mattemoore @tomdean22 has this worked for either of you before or is this a new setup and it doesn't work?
> 
> @mattemoore if you can maybe have a play around with the above code in `node_modules` to see if you can get something different would be super useful. You could start by removing the `if` clause and always provide `username` and `password` which happen only when you provide a token and see if that yields different results.

This is a new set up for us. Within the last couple weeks.

**Comment by mattemoore** [comment]
@benjdlambert I will have time this week to look into this.

Can you confirm the following workflow please?
1. Modify `code tc-backstage\node_modules\@backstage\plugin-scaffolder-backend-module-azure\dist\index.cjs.js`
2. Run `yarn dev`
3. Test/validate/debug etc.?

**Comment by benjdlambert** [comment]
Yep that should do it 👍

**Comment by awanlin** [comment]
Popping this on my todo list to look into, mostly as I might have some ideas to help with debugging this.

@sanderaernouts who added support for managed identity is on vacation but including him here if he happens to be around and has a suggestion.

**Comment by sanderaernouts** [comment]
@awanlin it’s a bit hard to follow up on this on my phone. But IIRC the underlying (nodejs) library sets the Authorization HTTP header when passing the credentials. Username password results in basic authentication and token should result in Bearer (token) authentication. But from the error message it seems that doesn’t happen. 401 generally means you did not provide any authentication as opposed to the wrong credentials.

It has been a while, but I believe I also added support for the token option in the (git) credential in the first place so maybe that is a good place to start? I think I added support for both a promise and a value so maybe something is off there.

I’d be happy to help out but I’m on holiday till the 26th.

**Comment by sanderaernouts** [comment]
@awanlin I think I found my mistake. I set the Authorization header but that should have been the Authentication header: https://github.com/backstage/backstage/blob/661b354277d71a9a5005930b6b900c4d24601965/plugins/scaffolder-node/src/scm/git.ts#L91. 

I’m a bit confused because I’m pretty sure the Authorization header is the correct one but the docs for isomorphic git say otherwise: https://github.com/isomorphic-git/isomorphic-git/blob/9103da548885b85dff89183d9ba654d23c6c3b09/docs/onAuth.md?plain=1#L17.

That being said, in hindsight I think using the onAuth callback instead of providing a “fixed” token is the more correct approach. You should be able to pass the bearer token as the value for password and the username name doesn’t matter as long as it’s not empty. The callback may then have a chance to refresh the token if needed whereas the fixed “token” may contain an expired token.

**Comment by awanlin** [comment]
Thanks for taking a look @sanderaernouts, this is more than I expected as I know you are away.

@mattemoore, can you perhaps play around with the ideas shared?

**Comment by mattemoore** [comment]
When using PAT `auth` is:
```
{
   username: 'notempty',
   password: 'REDACTED PAT'
}
```

When using app registration (same app registration used for azure authentication): 
azureDevOps:
  host: dev.azure.com
  token: dummy-token-value
  organization: transport-canada

```
integrations:
  azure:
    - host: dev.azure.com
      credentials:
        # - personalAccessToken: ${AZURE_DEVOPS_PAT}
        - clientId: ${AZURE_CLIENT_ID}
          clientSecret: ${AZURE_CLIENT_SECRET}
          tenantId: ${AZURE_TENANT_ID}
```
I get the following error: 
`TF401444: Please sign-in at least once as 2008ffa9-c9b2-4d97-9ad9-4ace25386be7\2008ffa9-c9b2-4d97-9ad9-4ace25386be7\7ff59517-7427-409f-8235-216331e94bac in a web browser to enable access to the service.`
![image](https://github.com/user-attachments/assets/34a082c6-a752-4635-870d-64046d5d870d)

I, unlike @tomdean22, cannot get the repo creation step to complete:

```
2024-09-09T18:16:53.000Z Beginning step Publish to Azure
[93](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-93)2024-09-09T18:16:53.000Z Error: TF401444: Please sign-in at least once as 2008ffa9-c9b2-4d97-9ad9-4ace25386be7\2008ffa9-c9b2-4d97-9ad9-4ace25386be7\7ff59517-7427-409f-8235-216331e94bac in a web browser to enable access to the service.
[94](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-94)    at RestClient.<anonymous> (C:\Users\MOOREMA\code\Backstage\Backstage\tc-backstage\node_modules\typed-rest-client\RestClient.js:202:31)
[95](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-95)    at Generator.next (<anonymous>)
[96](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-96)    at fulfilled (C:\Users\MOOREMA\code\Backstage\Backstage\tc-backstage\node_modules\typed-rest-client\RestClient.js:6:58)
[97](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-97)    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

The service account has been granted Create Repository rights explicitly.

**Comment by tomdean22** [comment]
> When using PAT `auth` is:
> 
> ```
> {
>    username: 'notempty',
>    password: 'REDACTED PAT'
> }
> ```
> 
> When using app registration (same app registration used for azure authentication): azureDevOps: host: dev.azure.com token: dummy-token-value organization: transport-canada
> 
> ```
> integrations:
>   azure:
>     - host: dev.azure.com
>       credentials:
>         # - personalAccessToken: ${AZURE_DEVOPS_PAT}
>         - clientId: ${AZURE_CLIENT_ID}
>           clientSecret: ${AZURE_CLIENT_SECRET}
>           tenantId: ${AZURE_TENANT_ID}
> ```
> 
> I get the following error: `TF401444: Please sign-in at least once as 2008ffa9-c9b2-4d97-9ad9-4ace25386be7\2008ffa9-c9b2-4d97-9ad9-4ace25386be7\7ff59517-7427-409f-8235-216331e94bac in a web browser to enable access to the service.` ![image](https://private-user-images.githubusercontent.com/8275624/365762773-34a082c6-a752-4635-870d-64046d5d870d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5MTIyMTEsIm5iZiI6MTcyNTkxMTkxMSwicGF0aCI6Ii84Mjc1NjI0LzM2NTc2Mjc3My0zNGEwODJjNi1hNzUyLTQ2MzUtODcwZC02NDA0NmQ1ZDg3MGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDlUMTk1ODMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmVkOTBmNjA0ODA2MzY3OWJhZDkxNzEwOGQ2NjFkM2M3MWRmODZiMTBhYTgyYzVjYTllMTYxZGNjN2NiNmFlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.u7fs-U7-kmg81oWVMfmnPyKJu2luGah7Ef8r03sGnVE)
> 
> I, unlike @tomdean22, cannot get the repo creation step to complete:
> 
> ```
> 2024-09-09T18:16:53.000Z Beginning step Publish to Azure
> [93](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-93)2024-09-09T18:16:53.000Z Error: TF401444: Please sign-in at least once as 2008ffa9-c9b2-4d97-9ad9-4ace25386be7\2008ffa9-c9b2-4d97-9ad9-4ace25386be7\7ff59517-7427-409f-8235-216331e94bac in a web browser to enable access to the service.
> [94](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-94)    at RestClient.<anonymous> (C:\Users\MOOREMA\code\Backstage\Backstage\tc-backstage\node_modules\typed-rest-client\RestClient.js:202:31)
> [95](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-95)    at Generator.next (<anonymous>)
> [96](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-96)    at fulfilled (C:\Users\MOOREMA\code\Backstage\Backstage\tc-backstage\node_modules\typed-rest-client\RestClient.js:6:58)
> [97](http://localhost:3000/create/tasks/3ce1e5e3-4f5b-44fe-a7d4-8cfabbc14460#line-97)    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> ```
> 
> The service account has been granted Create Repository rights explicitly.

I get this error when the account isn't a user in our Azure DevOps org. Add the app registration or managed identity to your Azure DevOps org as a user. Once I added it as a user, I put it in the **Project Collection Service Accounts** group (org level) to elevate its permissions, Project Collection Build Service Accounts might also do the trick.

**Comment by mattemoore** [comment]
I updated `tc-backstage\node_modules\@backstage\plugin-scaffolder-node\dist\index.cjs.js` from:
```
class Git {
  constructor(config) {
    this.config = config;
    this.onAuth = config.onAuth;
    this.headers = {
      "user-agent": "git/@isomorphic-git",
      ...config.token ? { Authorization: `Bearer ${config.token}` } : {}
    };
  }
```
to
```
class Git {
  constructor(config) {
    this.config = config;
    this.onAuth = config.onAuth;
    this.headers = {
      "user-agent": "git/@isomorphic-git",
      ...config.token ? { **Authentication**: `Bearer ${config.token}` } : {}
    };
    console.log(`--------------------------------------------${this.headers}`)
  }
 ```
as requested and still received the 401 error:
```
2024-09-10T12:05:42.000Z info: Pushing directory to remote {dir=C:\Users\MOOREMA\AppData\Local\Temp\97623fd2-87dc-4682-a259-67020b0e6f21,remote=origin}

(http://localhost:3000/create/tasks/97623fd2-87dc-4682-a259-67020b0e6f21#line-98)2024-09-10T12:05:42.000Z error: Failed to push to repo {dir=C:\Users\MOOREMA\AppData\Local\Temp\97623fd2-87dc-4682-a259-67020b0e6f21, remote=origin}

(http://localhost:3000/create/tasks/97623fd2-87dc-4682-a259-67020b0e6f21#line-99)2024-09-10T12:05:42.000Z Error: HTTP Error: 401 Unauthorized {data={"statusCode":401,"statusMessage":"Unauthorized","response":""}}

(http://localhost:3000/create/tasks/97623fd2-87dc-4682-a259-67020b0e6f21#line-100)    at Git.push (C:\Users\MOOREMA\code\Backstage\Backstage\tc-backstage\node_modules\@backstage\plugin-scaffolder-node\src\scm\git.ts:289:7)
```

**Comment by mattemoore** [comment]
@awanlin  and @sanderaernouts 
I can confirm that changing the code in https://github.com/backstage/backstage/commit/661b354277d71a9a5005930b6b900c4d24601965 to:
```
    const auth = {
        username: "notempty",
        password: ctx.input.token ?? credentials.token
      }
```
works with both PAT and Service Account.

**Comment by benjdlambert** [comment]
@mattemoore @sanderaernouts nice that we have a fix! 

@mattemoore do you want to raise a PR adjusting the logic to the working solution for both approaches? :pray:

**Comment by mattemoore** [comment]
@benjdlambert yes willing to do that.  Can you point me to the contribution process documentation.  I have not created a PR for backstage before.

**Comment by awanlin** [comment]
Hi @mattemoore, that's here: https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md. Feel free to ping here if you have issue or in the PR when it's ready. Happy to help and do early reviews to move things along 🚀

**Comment by mattemoore** [comment]
@awanlin I have blessing from legal to commit. I've forked and cloned and made my change - ready to test.  How does one pull down the community plugins?  Do I just install it normally as per the plugin docs?  If so what dir to I cd into first before installing plugin?

**Comment by awanlin** [comment]
This work would be done in the main Backstage repo, so the one you are in for this issue. This is NOT related to the Community Plugins repo. You do not need to install any plugins 👍 

The main Backstage repo has an existing example app that you can use, so `yarn install`, and then `yarn dev` to start it up. In this case I think all you should have to do is import a Software Template that will trigger the issue as well as setup the Azure integration config in your `app-config.local.yaml`.

**Comment by mattemoore** [comment]
@awanlin sorry I used the wrong words in my comment above.  I do realize I'm making a change to a plugin that comes with backstage (even though I mentioned Community Plugins :) ).  I'm in the right repo and have the example app running.  I'm all good there.

When I go to publish my template to Azure it fails with:
```
Error
NotFoundError

Message
Template action with ID 'publish:azure' is not registered.
```

**Comment by mattemoore** [comment]
@awanlin nvm, I forgot to add the azure scaffolder plugin to my backend in `packages\backend\src\index.ts`

**Comment by acierto** [comment]
It sounds to me that having some better error handling can save time for troubleshooting missing plugins in classpath

**Comment by awanlin** [comment]
All good @mattemoore, yeah, not all the Scaffolder modules are added to the example-app backend so in this case you would need to add it BUT make sure you use the `workspace:^` syntax when you add it to the `package.json` file so that way you use the version in the repo and not the one published to NPM. It would be a lot like this:

https://github.com/backstage/backstage/blob/77146ae7f199fc3334b16b719a102cf6f91075bc/packages/backend/package.json#L52C5-L52C73

**Comment by mattemoore** [comment]
@awanlin adding it to the backend was all that I needed.  I did not have to add it to `package.json`.

PR is up: https://github.com/backstage/backstage/pull/26661

**Comment by acierto** [comment]
As PR was merged, closing this bug report
